### PR TITLE
opentelemetrytracer: Dynatrace sampler: Enable adaptative sampling

### PR DESCRIPTION
**Pull request by samohte**
Commit Message: Dynatrace sampler to fetch configuration from an API

Additional Description: In [#32598](https://github.com/envoyproxy/envoy/pull/32598) a custom Dynatrace sampler was added to Envoy. This sampler should now be extended to fetch its configuration via an HTTP call from a Dyantrace cluster.
Testing: Unit, Integration, Manual
Docs Changes: N/A
Release Notes:Dynatrace sampler to fetch configuration from an API
Platform Specific Features: N/A
[Optional Runtime guard:] N/A
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc [hunk 1]
```diff
@@ -8,10 +10,98 @@ namespace Extensions {
...
+
+std::chrono::milliseconds getTimeout(envoy::config::core::v3::HttpUri& http_uri) {
+  auto timeout =
+      std::chrono::milliseconds(DurationUtil::durationToMilliseconds(http_uri.timeout()));
+  // we want to ensure that we don't have more than one pending request
+  if (timeout > (TIMER_INTERVAL / 2)) {
+    timeout = (TIMER_INTERVAL / 2);
+  }
+  return timeout;
+}
```
Comment (wbpcode) [inline source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:40]: does this necessary? Seems you will only re-enable the timer when the `onSucess` or `onFailure` is called. So, that means you will only emit a new request after last request is completed?

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc [hunk 2]
```diff
@@ -8,10 +10,98 @@ namespace Extensions {
...
+      ENVOY_LOG(warn, "Failed to parse sampling configuration received from Dynatrace: {}",
+                http_response->bodyAsString());
+    }
+  } else {
+    ENVOY_LOG(warn, "Failed to get sampling configuration from Dynatrace: {}", response_code);
+  }
+
+  if (json_valid || reEnableTimer(static_cast<Http::Code>(response_code))) {
+    timer_->enableTimer(std::chrono::seconds(TIMER_INTERVAL));
+  }
```
Comment (wbpcode) [inline source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:95]: Is it reasonable design not to enable the timer if the body is invalid or the response code is unexpected value?

And I think at least here should have an error log to tell user the timer will never be enabled again.

**Review by wbpcode — COMMENTED** [review]
Thanks for you contribution. And some comments are added. :)

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc (hunk 2 above)
Comment (samohte) [reply source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:95]: Response codes lie 400, 401. 403, 404, ... are caused by a config error (wrong uri, wrong/invalid token). Therefore we thought the re-enabling the timer does not make sense.
I agree, we should log an error if we never enable the timer.

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc (hunk 1 above)
Comment (samohte) [reply source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc]: You are right, it is not possible to have more than one panding request. I have removed the adjustment in 535d99b7932c49f082bf5e8e1fda969708de1d32

**Comment by samohte** [comment]
> Thanks for you contribution. And some comments are added. :)

Thanks for your valuable feedback. I addressed your comments in 535d99b7932c49f082bf5e8e1fda969708de1d32

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc [hunk 3]
```diff
@@ -8,10 +10,92 @@ namespace Extensions {
...
+
+void SamplerConfigProviderImpl::onSuccess(const Http::AsyncClient::Request& /*request*/,
+                                          Http::ResponseMessagePtr&& http_response) {
+  active_request_ = nullptr;
+  const auto response_code = Http::Utility::getResponseStatus(http_response->headers());
+  bool json_valid = false;
+  if (response_code == enumToInt(Http::Code::OK)) {
+    ENVOY_LOG(debug, "Received sampling configuration from Dynatrace: {}",
+              http_response->bodyAsString());
+    json_valid = sampler_config_.parse(http_response->bodyAsString());
```
Comment (wbpcode) [inline source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:76]: Will we access the `sampler_config_` at worker threads? If yes, then seems there maybe a contention.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for you in timely update. Now, LGTM overall except the last comment about the thread safe. And please resolve the conflict. Thanks.

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc (hunk 3 above)
Comment (samohte) [reply source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:76]: yes, `sampler_config` is used in worker threads. But it has only two members:
* one is `const` and never modified during runtime.
* the other one is set during `parse` and read in the worker threads. It is an `std::atomic<uint32_t>`, so it should be ok to access the variable in multiple threads.

**Comment by samohte** [comment]
> Thanks for you in timely update. Now, LGTM overall except the last comment about the thread safe. And please resolve the conflict. Thanks.

I have resolved the conflict and answered your comment. Pls. let me know what you think about.

**Review by wbpcode — APPROVED** [review]

**Pull request merged**